simd/unit_tests/TestSIMD.cpp

```diff
@@ -0,0 +1,376 @@
+  template <class T>
+  KOKKOS_INLINE_FUNCTION auto on_device(T const& a, T const& b) const {
+    return a * b;
+  }
+};
+
+class divides {
+ public:
+  template <class T>
+  auto on_host(T const& a, T const& b) const {
+    return a / b;
+  }
+  template <class T>
+  KOKKOS_INLINE_FUNCTION auto on_device(T const& a, T const& b) const {
+    return a / b;
+  }
+};
+
+template <class Abi>
+inline void host_check_math_ops() {
+  std::size_t constexpr n     = 11;
+  double const first_args[n]  = {1, 2, -1, 10, 0, 1, -2, 10, 0, 1, -2};
+  double const second_args[n] = {1, 2, 1, 1, 1, -3, -2, 1, 13, -3, -2};
+  host_check_binary_op_all_loaders<Abi>(plus(), n, first_args, second_args);
+  host_check_binary_op_all_loaders<Abi>(minus(), n, first_args, second_args);
+  host_check_binary_op_all_loaders<Abi>(multiplies(), n, first_args,
+                                        second_args);
+  host_check_binary_op_all_loaders<Abi>(divides(), n, first_args, second_args);
+}
+
+template <class Abi>
+KOKKOS_INLINE_FUNCTION void device_check_math_ops() {
+  std::size_t constexpr n     = 11;
+  double const first_args[n]  = {1, 2, -1, 10, 0, 1, -2, 10, 0, 1, -2};
+  double const second_args[n] = {1, 2, 1, 1, 1, -3, -2, 1, 13, -3, -2};
+  device_check_binary_op_all_loaders<Abi>(plus(), n, first_args, second_args);
+  device_check_binary_op_all_loaders<Abi>(minus(), n, first_args, second_args);
+  device_check_binary_op_all_loaders<Abi>(multiplies(), n, first_args,
+                                          second_args);
+  device_check_binary_op_all_loaders<Abi>(divides(), n, first_args,
+                                          second_args);
+}
+
+template <class Abi>
+inline void host_check_abi() {
+  host_check_math_ops<Abi>();
+}
+
+template <class Abi>
+KOKKOS_INLINE_FUNCTION void device_check_abi() {
+  device_check_math_ops<Abi>();
+}
+
+inline void host_check_abis(Kokkos::Experimental::simd_abi::abi_set<>) {}
+
+KOKKOS_INLINE_FUNCTION void device_check_abis(
+    Kokkos::Experimental::simd_abi::abi_set<>) {}
+
+template <class FirstAbi, class... RestAbis>
+inline void host_check_abis(
+    Kokkos::Experimental::simd_abi::abi_set<FirstAbi, RestAbis...>) {
+  host_check_abi<FirstAbi>();
+  host_check_abis(Kokkos::Experimental::simd_abi::abi_set<RestAbis...>());
+}
+
+template <class FirstAbi, class... RestAbis>
+KOKKOS_INLINE_FUNCTION void device_check_abis(
+    Kokkos::Experimental::simd_abi::abi_set<FirstAbi, RestAbis...>) {
+  device_check_abi<FirstAbi>();
+  device_check_abis(Kokkos::Experimental::simd_abi::abi_set<RestAbis...>());
+}
+
+TEST(simd, host) {
+  host_check_abis(Kokkos::Experimental::simd_abi::host_abi_set());
+}
+
+class simd_device_functor {
+ public:
+  KOKKOS_INLINE_FUNCTION void operator()(int) const {
+    device_check_abis(Kokkos::Experimental::simd_abi::device_abi_set());
+  }
+};
+
+TEST(simd, device) {
+  Kokkos::parallel_for(Kokkos::RangePolicy<Kokkos::IndexType<int>>(0, 1),
+                       simd_device_functor());
+}
```

Comment:
I only skimmed the implementation files, but I think they largely look fine.

I've looked more at the tests since those are the best way to ensure we have all the desired functionality as we add backends etc. Here's what I see missing right now:
1) The compound assignment operators (+=,-=, *=,/=)
2) Comparison operators
3) All the math functions (max/min/abs/acos/asin/atan/cos/exp/log/pow/sin/sqrt/tan/tanh are the ones I know we use).
4) All of the above where one or more of the operands are non-SIMD types.

I don't feel strongly about including them in this PR or as follow-up.

Reply:
Thanks Victor! Most of that is there but not covered in tests, but what is not there is mainly the full `<cmath>` functions. Maybe given the slowness of Kokkos PR review I will add them.